apps/emqx_machine/src/emqx_cluster_rpc.erl

```diff
@@ -0,0 +1,325 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_cluster_rpc).
+-behaviour(gen_statem).
+
+%% API
+-export([start_link/0, mnesia/1]).
+-export([multicall/3, multicall/4, query/1, reset/0, status/0]).
+
+-export([init/1, format_status/2, handle_event/4, terminate/3,
+    code_change/4, callback_mode/0]).
+
+-ifdef(TEST).
+-compile(export_all).
+-compile(nowarn_export_all).
+-export([start_link/3]).
+-endif.
+
+-boot_mnesia({mnesia, [boot]}).
+-copy_mnesia({mnesia, [copy]}).
+
+-include_lib("emqx/include/emqx.hrl").
+-include_lib("emqx/include/logger.hrl").
+-include("emqx_cluster_rpc.hrl").
+
+-rlog_shard({?COMMON_SHARD, ?CLUSTER_MFA}).
+-rlog_shard({?COMMON_SHARD, ?CLUSTER_COMMIT}).
+
+-define(CATCH_UP, catch_up).
+-define(REALTIME, realtime).
+-define(CATCH_UP_AFTER(_Ms_), {state_timeout, _Ms_, catch_up_delay}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+mnesia(boot) ->
+    ok = ekka_mnesia:create_table(?CLUSTER_MFA, [
+        {type, ordered_set},
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, cluster_rpc_mfa},
+        {attributes, record_info(fields, cluster_rpc_mfa)}]),
+    ok = ekka_mnesia:create_table(?CLUSTER_COMMIT, [
+        {type, set},
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, cluster_rpc_commit},
+        {attributes, record_info(fields, cluster_rpc_commit)}]);
+mnesia(copy) ->
+    ok = ekka_mnesia:copy_table(cluster_rpc_mfa, disc_copies),
+    ok = ekka_mnesia:copy_table(cluster_rpc_commit, disc_copies).
+
+start_link() ->
+    RetryMs = application:get_env(emqx_machine, cluster_call_retry_interval, 1000),
+    start_link(node(), ?MODULE, RetryMs).
+start_link(Node, Name, RetryMs) ->
+    gen_statem:start_link({local, Name}, ?MODULE, [Node, RetryMs], []).
+
+-spec multicall(Module, Function, Args) -> {ok, TnxId} | {error, Reason} when
+    Module :: module(),
+    Function :: atom(),
+    Args :: [term()],
+    TnxId :: pos_integer(),
+    Reason :: term().
```

Comment:
Coding style: you could omit specifying reason type explicitly, if you change the return type to `| {error, _Reason}`. Dialyzer will assume that it's a `term()`.